### PR TITLE
Add ExitOutsideMain rule

### PR DESCRIPTION
**Pull request by 3flex**
A port of [Error Prone's SystemExitOutsideMain rule](http://errorprone.info/bugpattern/SystemExitOutsideMain).

The rule also flags usage of `System.exit` or `exitProcess` that are in `main` but in a nested function. It's debatable whether it makes sense to flag usage in the nested function since it's technically in the `main` function and maybe not really an issue. I'm looking for feedback on that. If the logic stays as originally submitted then I'll try and come up with a better name for the rule.

### CODE REVIEW

File: detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ExitOutsideMain.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ *     exitProcess(2)
+ *   } else {
+ *     exitProcess(0)
+ *   }
+ * }
+ * </compliant>
+ *
+ * @requiresTypeResolution
+ */
+
```
Comment (cortinico) [inline detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ExitOutsideMain.kt:47]: This white line is unnecessary

File: detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ExitOutsideMain.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    override fun visitCallExpression(expression: KtCallExpression) {
+        if (context == BindingContext.EMPTY) return
+
+        if (expression.getStrictParentOfType<KtNamedFunction>()?.isMainFunction() == true) return
+        val fqName = expression.getResolvedCall(bindingContext)?.resultingDescriptor?.fqNameOrNull() ?: return
+
+        when (fqName.asString()) {
+            "kotlin.system.exitProcess" -> { report(CodeSmell(issue, Entity.from(expression), issue.description)) }
+            "java.lang.System.exit" -> { report(CodeSmell(issue, Entity.from(expression), issue.description)) }
```
Comment (cortinico) [inline detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ExitOutsideMain.kt:65]:
```suggestion
            "kotlin.system.exitProcess", "java.lang.System.exit" -> { report(CodeSmell(issue, Entity.from(expression), issue.description)) }
```

File: detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ExitOutsideMain.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+        super.visitCallExpression(expression)
```
Comment (cortinico) [inline detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ExitOutsideMain.kt:68]: This should be moved as first statement

**Review by cortinico — APPROVED** [review]

**Review by schalkms — COMMENTED** [review]
This new rule looks good in my opinion. Is there any chance that the next detekt version includes this rule too?

**Review by BraisGabin — APPROVED** [review]
I agree that nested functions should not be flagged. But it seems like a really edge case to me. I'm OK with this or changing the rule to allow those cases, up to you.

**Comment by 3flex** [comment]
Thanks for the feedback and sorry for the turnaround time. This should be good to go. I've left the logic as is for the nested functions as I agree it's very much an edge case and IMHO not something you'd want to allow anyway.

**Pull request merged**